### ISSUE TITLE
Code Coverage

### ISSUE THREAD

**Issue by AllanOricil** [apex-enterprise-patterns/fflib-apex-common#239]
I tried to deploy the fflib-apex-common classes and mocks but the code coverage was not enough.
![image](https://user-images.githubusercontent.com/55927613/66050480-34915080-e525-11e9-802f-07fe1518ec73.png)

**Comment by AllanOricil** [comment]
I was able to deploy it when I create a class for the fflib_QUalified MEthodAndArgValues.
But the other two classes still need coverage to be deployed.

![image](https://user-images.githubusercontent.com/55927613/66054777-6823a900-e52c-11e9-847b-311cde6aae2b.png)

**Comment by cpfrommer** [comment]
I saw the same coverage issue with the fflib_SObjectUnitOfWork and fflib_SObjectSelector classes.  I saw some comments relating to #212 about how the additions can't really be unit tested, so that might explain the low coverage on fflib_SObjectUnitOfWork.  Not sure about fflib_SObjectSelector.

**Comment by cpfrommer** [comment]
So I think between the addition of the registerPublish* methods added in pull request #2 (which currently have no coverage) and the new untestable stuff in #212 the coverage on fflib_SObjectUnitOfWork slipped below 75%.

I might try to fix this myself and submit a PR at some point...

**Comment by evanmcd** [comment]
@cpfrommer Hey, I don't suppose you ever got to writing additional tests for this, did you? I'm hitting the same issue now.

Thanks.

**Comment by cpfrommer** [comment]
@evanmcd no, unfortunately I haven't gotten to that :(

**Comment by evanmcd** [comment]
OK. Thx for the quick update. I put in a PR for a missing test class on the mocks library earlier today. Maybe I'll be able to get to this at some point as well.

**Comment by MorganMarchese** [comment]
8/23/2021 - this is still an issue blocking "Run Specified Tests" deployments.

**Comment by ImJohnMDaniel** [comment]
@MorganMarchese -- which classes are you seeing insufficient when you run specified tests??

**Comment by MorganMarchese** [comment]
@ImJohnMDaniel 

fflib_SObjectSelector and fflib_SObjectUnitOfWork both have under 75% coverage. Since I need fflib for an upcoming prod deployment, I've been working on my own personal coverage improvements since my comment yesterday - but if I recall correctly, the original count was something like...

fflib_SObjectSelector: 48%
fflib_SObjectUnitofWork: 68%

**Comment by ImJohnMDaniel** [comment]
OK.  This is somewhat of a known issue.   It doesn't affect most folks because they either deploy the entire framework or the entire app and run all of the unit tests instead these specific ones.  The other option that is getting more traction with folks is make a private clone of this repo, add your own second generation unlocked package info to the `sfdx-project.json` and deploy the framework as an unlocked package.  That actually shifts when the unit tests are run from "deployment time" back to "package version creation time."

What is driving the need to specify these particular unit tests during the deployment?

**Comment by MorganMarchese** [comment]
@ImJohnMDaniel unfortunately what is driving the need is a bad org. I recently adopted an org where you **cannot** run/pass all test classes during a production deployment due to 5-7 years of mismanaged consultant code being rammed into production.

I've been brought on to resolve some core architectural/process challenges within the org, but as of today due to the number of failing test classes (hundreds) the only way I can deploy anything is to build strong unit tests for all new/changed classes which independently cover the code at 75%+.

One of my upcoming improvements is the introduction of the Selector layer pattern from FFLib, but I can't deploy FFLib until I get that coverage above 75% for each class. I managed to get UnitOfWork up to 82% by adding custom coverage for some things that can't be covered by standard object unit tests (like registering with External Ids, and Event Publishing methods), and working on getting above 75% with SObjectSelector today. So far I'm at 62%.

I haven't worked much with unlocked packages yet. By doing it this way, are you saying that Salesforce doesn't enforce 75% coverage on each class during package creation time?

**Comment by AllanOricil** [comment]
@ImJohnMDaniel im my case it was a company policy and architect's agreement to not allow libraries that are not covered. I think this is going to be a common scenario that would imped developers to try it in prod. 
So, it it is really not possible to get the required code coverage, add a section on the readme where you explain why it has less than 75% coverage so architects are aware of it before deploying.

**Comment by MorganMarchese** [comment]
Is there any reason (other than that nobody has done it yet) that the main fflib-apex-common hasn't been packaged/offered as an unlocked package?

**Comment by PeterLinBiz** [comment]
@MorganMarchese,
Each SFDX-UP version is managed and (100% controlled) by a specific Salesforce Dev Hub Org (regular production org, Developer Edition org or Salesforce Partner Business Org, etc.)

If I were to share with you my SFDX-UP version, you probably don't want to use it to install into any of your production org, because I can change that SFDX-UP version any minute (so it may only work in my production org but NOT work in any other production orgs, e.g. one of your production orgs) which is out of your control.

So, I suggest that each production org (Dev Hub Org) should manage its own SFDX-UP Packages and versions with 100% control over the changes.

**Comment by PeterLinBiz** [comment]
@MorganMarchese,
Each SFDX-UP version is managed and (100% controlled) by a specific Salesforce Dev Hub Org (Regular Production Org, Developer Edition Org or Salesforce Partner Business Org, etc.).

If I were to share with you my SFDX-UP version, you will probably don't want it, because I cannot change it any minute which is out of your control.

**Comment by fransf-wtax** [comment]
Hi all, running into the same issues: insufficient code coverage for `fflib_SObjectSelector` and `fflib_SObjectUnitOfWork`. I've already forked the project so the quickest way out for me is probably just to add some tests that at least work in my orgs and that achieve the required coverage (or to go the unlocked package route suggested). But it does feel like a waste for every user of this great library to reinvent the wheel here.

- Why not release fflib-apex-common as a first generation managed package? That wouldn't have the dev hub coupling that unlocked packages have.

**Comment by daveespo** [comment]
1GP managed packages are no better, really. First off, it would need to go through security review which is only available to Partners -- we're just a bunch of guys maintaining a project with no business entity behind it.

Second, there is no support on the platform for installing multiple packages in one install via AppExchange install -- so if anyone wanted to integrate the 1GP into their ISV product, it would preclude it from being installed via the 'Get it now' feature. No ISV would limit themselves in such a way so they'd likely vendor the fflib code (like they do today) .. but that brings us to point three:

Third, we'd need to expose every class and method as `global` so it could be accessed outside of the 1GP package. That would basically freeze the interface and limit future refactoring options dramatically. Again, not something we're going to do. 

In short, there is sufficient test coverage in this project if you deploy it in its entirety. Trying to do piecemeal deployments isn't possible and it's covered in earlier comments on this thread (notably https://github.com/apex-enterprise-patterns/fflib-apex-common/issues/239#issuecomment-904646020 and the ones after it)

**Comment by fransf-wtax** [comment]
Thanks for that explanation @daveespo :)

I've managed to get the coverage to over 75% on my fork for the two classes where it was lacking. Is it worth the effort to submit a pull request? They're probably not the best tests, I just looked at which lines were not covered and were not related to multi-currency, and wrote the smallest unit test I could to get them covered. But it might help others in the same position as me.

**Comment by MorganMarchese** [comment]
> 1GP managed packages are no better, really. First off, it would need to go through security review which is only available to Partners -- we're just a bunch of guys maintaining a project with no business entity behind it.
> 
> Second, there is no support on the platform for installing multiple packages in one install via AppExchange install -- so if anyone wanted to integrate the 1GP into their ISV product, it would preclude it from being installed via the 'Get it now' feature. No ISV would limit themselves in such a way so they'd likely vendor the fflib code (like they do today) .. but that brings us to point three:
> 
> Third, we'd need to expose every class and method as `global` so it could be accessed outside of the 1GP package. That would basically freeze the interface and limit future refactoring options dramatically. Again, not something we're going to do.
> 
> In short, there is sufficient test coverage in this project if you deploy it in its entirety. Trying to do piecemeal deployments isn't possible and it's covered in earlier comments on this thread (notably [#239 (comment)](https://github.com/apex-enterprise-patterns/fflib-apex-common/issues/239#issuecomment-904646020) and the ones after it)

To be clear @daveespo , my use case is not to do "piecemeal" deployments of small portions of fflib. The real challenge is the inability to install it into an org which has even a single failing test class in production. In the org I joined last year, we are forced to do "run specified tests" deployments at present.

Taking aside the obvious underpinning issue which is that our tests shouldn't be failing in production (which I'm working on), the fflib library cannot be deployed to orgs in this state because when you use "Run Specified Tests" Salesforce requires that EACH class being deployed has a minimum of 75% coverage, rather than the collective whole being > 75%. 

This makes it an increasingly difficult challenge for people to adopt this library as a mechanism for assisting in fixing bad orgs. Thankfully I was able to overcome this by taking the earlier advice of creating two 2GP's for ApexMocks and Common,

@fransf-wtax it was my experience that coverage couldn't be "generically" increased above 75%, which is why these classes lack coverage. If you've somehow managed to create non-org dependent test classes that give adequate coverage for these, they'd probably welcome the contribution.

**Comment by fransf-wtax** [comment]
I just used the standard SObjects they were already testing with -- `Opportunity` and `Product2` -- so as far as I can tell the tests are not org-dependent. I'll isolate my changes into a single commit and submit a pull request. Let's see what happens :-)

**Comment by fransf-wtax** [comment]
https://github.com/apex-enterprise-patterns/fflib-apex-common/pull/366
